File: enode/src/main/java/org/enodeframework/queue/publishableexceptions/DefaultPublishableExceptionPublisher.java
package org.enodeframework.queue.publishableexceptions;

import org.enodeframework.common.serializing.SerializeService;
import org.enodeframework.common.utils.Assert;
import org.enodeframework.domain.DomainExceptionMessage;
import org.enodeframework.infrastructure.TypeNameProvider;
import org.enodeframework.messaging.MessagePublisher;
import org.enodeframework.queue.MessageTypeCode;
import org.enodeframework.queue.QueueMessage;
import org.enodeframework.queue.SendMessageService;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class DefaultPublishableExceptionPublisher implements MessagePublisher<DomainExceptionMessage> {

    private final String topic;

    private final String tag;

    private final SendMessageService sendMessageService;

    private final SerializeService serializeService;

    private final TypeNameProvider typeNameProvider;

    public DefaultPublishableExceptionPublisher(String topic, String tag, SendMessageService sendMessageService, SerializeService serializeService, TypeNameProvider typeNameProvider) {
        this.topic = topic;
        this.tag = tag;
        this.sendMessageService = sendMessageService;
        this.serializeService = serializeService;
        this.typeNameProvider = typeNameProvider;
    }

    protected QueueMessage createExceptionMessage(DomainExceptionMessage exception) {
        Assert.nonNull(topic, "topic");
        Map<String, Object> serializableInfo = new HashMap<>();
        exception.serializeTo(serializableInfo);
        GenericPublishableExceptionMessage exceptionMessage = new GenericPublishableExceptionMessage();
        exceptionMessage.setUniqueId(exception.getId());
        exceptionMessage.setExceptionType(typeNameProvider.getTypeName(exception.getClass()));
        exceptionMessage.setTimestamp(exception.getTimestamp());
        exceptionMessage.setSerializableInfo(serializableInfo);
        exceptionMessage.setItems(exception.getItems());
        String data = serializeService.serialize(exceptionMessage);
        String routeKey = exception.getId();
        QueueMessage queueMessage = new QueueMessage();
        queueMessage.setTopic(topic);
        queueMessage.setTag(tag);
        queueMessage.setBody(data);
        queueMessage.setRouteKey(routeKey);
        queueMessage.setType(MessageTypeCode.ExceptionMessage.getValue());
        queueMessage.setKey(exceptionMessage.getUniqueId());
        return queueMessage;
    }

    @Override
    public CompletableFuture<Boolean> publishAsync(DomainExceptionMessage message) {
        return sendMessageService.sendMessageAsync(createExceptionMessage(message));
    }
}

File: enode/src/main/java/org/enodeframework/queue/publishableexceptions/DefaultPublishableExceptionPublisher.kt
package org.enodeframework.queue.publishableexceptions

import org.enodeframework.common.serializing.SerializeService
import org.enodeframework.common.utils.Assert
import org.enodeframework.domain.DomainExceptionMessage
import org.enodeframework.infrastructure.TypeNameProvider
import org.enodeframework.messaging.MessagePublisher
import org.enodeframework.queue.MessageTypeCode
import org.enodeframework.queue.QueueMessage
import org.enodeframework.queue.SendMessageService
import java.util.concurrent.CompletableFuture

class DefaultPublishableExceptionPublisher(
    private val topic: String,
    private val tag: String,
    private val sendMessageService: SendMessageService,
    private val serializeService: SerializeService,
    private val typeNameProvider: TypeNameProvider
) : MessagePublisher<DomainExceptionMessage> {
    protected fun createExceptionMessage(exception: DomainExceptionMessage): QueueMessage {
        Assert.nonNull(topic, "topic")
        val serializableInfo: MutableMap<String, Any> = HashMap()
        exception.serializeTo(serializableInfo)
        val exceptionMessage = GenericPublishableExceptionMessage()
        exceptionMessage.uniqueId = exception.id
        exceptionMessage.exceptionType = typeNameProvider.getTypeName(exception.javaClass)
        exceptionMessage.timestamp = exception.timestamp
        exceptionMessage.serializableInfo = serializableInfo
        exceptionMessage.items = exception.items
        val data = serializeService.serialize(exceptionMessage)
        val routeKey = exception.id
        val queueMessage = QueueMessage()
        queueMessage.topic = topic
        queueMessage.tag = tag
        queueMessage.body = data
        queueMessage.routeKey = routeKey
        queueMessage.type = MessageTypeCode.ExceptionMessage.value
        queueMessage.key = exceptionMessage.uniqueId
        return queueMessage
    }

    override fun publishAsync(message: DomainExceptionMessage): CompletableFuture<Boolean> {
        return sendMessageService.sendMessageAsync(createExceptionMessage(message))
    }
}
